src/Transports.AspNetCore/WebSockets/SubscriptionList.cs

```diff
@@ -0,0 +1,228 @@
+namespace GraphQL.Server.Transports.AspNetCore.WebSockets;
+
+/// <summary>
+/// A list of subscriptions represented by <see cref="IDisposable"/> objects.
+/// All members are thread-safe.
+/// <br/><br/>
+/// Upon signalling the cancellation token, all future calls to any method except
+/// <see cref="Dispose"/> will throw an <see cref="OperationCanceledException"/>.
+/// <br/><br/>
+/// It is recommended to call <see cref="Dispose"/> immediately after signalling
+/// the cancellation token.
+/// </summary>
+public sealed class SubscriptionList : IDisposable
+{
+    private readonly CancellationToken _cancellationToken;
+    private readonly Dictionary<string, IDisposable> _subscriptions = new();
+    private readonly object _lock = new();
+
+    /// <summary>
+    /// Initializes a new instance.
+    /// </summary>
+    public SubscriptionList(CancellationToken cancellationToken)
+    {
+        _cancellationToken = cancellationToken;
+    }
+
+    /// <summary>
+    /// Disposes of all active subscriptions.
+    /// </summary>
+    public void Dispose()
+    {
+        List<IDisposable> subscriptionsToDispose;
+        lock (_lock)
+        {
+            subscriptionsToDispose = _subscriptions.Values.ToList();
+            _subscriptions.Clear();
+        }
+        foreach (var disposer in subscriptionsToDispose)
+        {
+            disposer.Dispose();
+        }
+    }
+
+    /// <summary>
+    /// Adds a subscription to the internal list if none is already in the list with the same id.
+    /// </summary>
+    /// <exception cref="ArgumentNullException"/>
+    /// <exception cref="OperationCanceledException"/>
+    public bool TryAdd(string id, IDisposable subscription)
+    {
+        if (id == null)
+            throw new ArgumentNullException(nameof(id));
+        if (subscription == null)
+            throw new ArgumentNullException(nameof(subscription));
+
+        lock (_lock)
+        {
+            _cancellationToken.ThrowIfCancellationRequested();
```

Comment:
Should this line be under lock statement? The same question for all other methods.

Reply:
It is EXCEEDINGLY important that all subscriptions are disposed of, since any not disposed of are GUARANTEED memory leaks. The GC will not collect lost objects. As such, most likely I wrote it this way to be sure the list was not shut down when adding or modifying entries. You can be sure I considered the question.

Typically a check like this might go on both sides of the lock, but here the redundant lock is unlikely to occur, so I implemented only the required check.

Reply:
I reviewed all these conditions again, and added disposed checks.

Reply:
I spent a few more hours on this, trying to determine if the cancellation token was necessary at all within this class.  I determined that it was not.  I made adjustments within GraphQL.AspNetCore3 including changes to tests to verify that everything works properly without the cancellation token.  Then I made the changes here.  So now this class has a very typical disposal pattern.

See PR:
- https://github.com/Shane32/GraphQL.AspNetCore3/pull/34